qa/common/util.py

```diff
@@ -0,0 +1,127 @@
+                # Extract key-value pairs from last matching row
+                (key, val) = key_val_pair[i].split(": ")
+                if key == "\"total\"":
+                    total = int(val)
+                elif key == "\"errors\"":
+                    errors = int(val)
+                elif key == "\"failures\"":
+                    failures = int(val)
+
+    if total == expected_number_of_test and errors == 0 and failures == 0:
+        f.close()
+        return True
+    
+    f.write("\n***\n*** Test Failed: Expected " + str(expected_number_of_test) + " test(s), " + str(total) + " test(s) executed, " + str(errors) + " test(s) had error, and " + str(failures) + " test(s) failed. \n***")
+    
+    f.close()
+    return False
+    """
+
+
+def check_csv_table_row_column(csv_file, expected_num_columns,
+                               expected_num_rows, tag):
+    try:
+        subprocess.run([
+            "bash", "-c",
+            "source " + util_sh_path + "; check_csv_table_row_column \"" +
+            str(csv_file) + "\" \"" + str(expected_num_columns) + "\" \"" +
+            str(expected_num_rows) + "\" \"" + str(tag) + "\""
+        ],
+                       check=True)
+    except subprocess.CalledProcessError:
+        print("\n***\n*** Test Output Verification Failed for " +
+              os.path.basename(csv_file) + ".\n***")
+        return False
+    return True
+
+
+def get_number_of_rows_logfile(log_file):
+    try:
+        subprocess.run([
+            "bash", "-c", "source " + util_sh_path +
+            "; get_number_of_rows_logfile \"" + str(log_file) + "\""
+        ],
+                       check=True)
+    except subprocess.CalledProcessError:
+        print("\n***\n*** Test Output Verification Failed for " +
+              os.path.basename(log_file) + ".\n***")
+        return False
+    return True
+
+
+def get_number_of_rows_csv(csv_file):
+    try:
+        subprocess.run([
+            "bash", "-c", "source " + util_sh_path +
+            "; get_number_of_rows_csv \"" + str(csv_file) + "\""
+        ],
+                       check=True)
+    except subprocess.CalledProcessError:
+        print("\n***\n*** Test Output Verification Failed for " +
+              os.path.basename(csv_file) + ".\n***")
+        return False
+    return True
+
+
+def check_log_table_row_column(log_file, expected_num_columns,
+                               expected_num_rows, tag):
+    try:
+        subprocess.run([
+            "bash", "-c",
+            "source " + util_sh_path + "; check_log_table_row_column \"" +
+            str(log_file) + "\" \"" + str(expected_num_columns) + "\" \"" +
+            str(expected_num_rows) + "\" \"" + str(tag) + "\""
+        ],
+                       check=True)
+    except subprocess.CalledProcessError:
+        print("\n***\n*** Test Output Verification Failed for " +
+              os.path.basename(log_file) + ".\n***")
+        return False
+    return True
```

Comment:
Can you add another function here that looks like below:

```
def check_test_results(log_file, expected_num_columns_inference, expected_num_columns_gpu, expected_num_columns_server, expected_num_rows_inference, expected_num_rows_server, expected_num_rows_gpu, gpu_filename_csv, inference_filename_csv, server_filename_csv):
```

This function will use the other functions that are present in this file to make sure that the results are correct. In all other tests, instead of calling `check_csv_table_row_column` and `check_log_table_row_column` separately this new function should be called.

Reply:
Can you please elaborate a bit more on the commonality you see between each L0 tests that uses `check_csv_table_row_column` and/or `check_log_table_row_column` ?

It's because each tests are somewhat distinct from each other. For example, the `L0_export_metrics/check_results.py` calls 
```
check_csv_table_row_column(
    os.path.abspath(self._model_metrics_GPU_file), metrics_num_columns,
    metrics_num_rows, output_tag)
```
while `L0_config/check_results.py` calls 
```
check_csv_table_row_column(
    os.path.abspath(self._model_metrics_GPU_file), metrics_num_columns,
    self._test_output_num_rows * self._num_gpus, output_tag)
```
They are both calling the same `check_csv_table_row_column` function on `model_metrics_GPU_file`, but notice the 3rd parameter is different. Thus, it's unlikely that those tests calling `check_csv_table_row_column` and/or `check_log_table_row_column` can be unified into one function without moving all the logics from each respective L0 tests into the single function. If that's the case, maybe it's better to simply leave the logic in their respective L0 tests?

Reply:
@kthui The third argument is taken care of in the `expected_num_columns_gpu` in the `check_results` function that I posted above. Each of these tests will provide a different value for that argument in the check_results. For example, `L0_export_metrics` will use `metrics_num_rows` for `expected_num_columns_gpu` argument and `self._test_output_num_rows * self._num_gpus` when the `check_results` is being used in `L0_config`. So, `check_results` will make sure that both the logs and CSV files have the expected number of rows and columns in each of the tests.

Does it make sense?

Reply:
I see, so you are just "combining" `check_csv_table_row_column` and `check_log_table_row_column` functions. Codes like 
```
        passed = check_log_table_row_column(...)
        passed &= check_csv_table_row_column(...)
        return passed
```
will become
```
        passed = check_results(...)
        passed &= check_results(...)
        return passed
```

Reply:
@kthui Yes, exactly. Also, I think you don't need two calls to `check_results`. It will be a single call:

```
passed = check_results(...)
return passed.
```

Reply:
I see.  I guess the fix is like you initially proposed that a lot of the `check_results.py`, those involving `check_csv_table_row_column` and `check_log_table_row_column`, should be refactored into `qa/common`, and variables like
```
        metrics_num_columns = 11
        inference_num_columns = 10
        server_metrics_num_columns = 5
        output_tag = "Model"
```
should remain in the `test.sh` where they originally came from.
Also their respective `test_config_generator.py`, since there are no different cases to generate, should remain in their original `config_generator.py` form, if configs need to be generated.

Reply:
Exactly, except that the `output_tag` should also be put into `qa/common`.  The `check_results` function will check all the model analyzer tables and CSV files and use the appropriate tag to find them. 

> Also their respective test_config_generator.py, since there are no different cases to generate, should remain in their original config_generator.py form, if configs need to be generated.

The `test_config_generator` changes look good to me except the tests that don't require it (like L0_doa and L0_unit_tests).

Reply:
I found an easier way to implement this. Since it's only a single function, it can be written as bash inside `qa/common/util.sh`. Then, the python part can be eliminated, and allow the bash scripts to simply call each other.